Refund by items: Refund items list - Step 2

0nko opened the pull request:
This PR adds the a new refund-by-items Fragment, layout and the item list implementation with all the necessary infrastructure. It was not possible to break this down further because there are some SavedState merge conflict resolutions that needed to be included.

This PR updates minor versions of a couple of dependencies -- most notably the Navigation component. This makes it possible to use the new `navGraphViewModels()` ViewModel scope, which ties the VM lifecycle to that of a subgraph. You can see it in action [here](https://github.com/woocommerce/woocommerce-android/pull/1622/files#diff-ac7fab20c439e6a4a5d5f02438e2d032R35). The SavedState injections [needed to be updated](https://github.com/woocommerce/woocommerce-android/pull/1622/files#diff-5f46e3b87cfb66d0983de260999194f1R28) to reflect this.

On the bright side, it's now possible to test the number picker dialog by tapping on the item count view. The refund by items functionality is not working, but you can select the number of items to refund and the refund subtotal and taxes are calculated. The `Select all` button works, as well.

By the way, I'm not sure what's going on with `ciklint` failing (`ktlint` passes on my machine), but the Step 3 PR is passing so just ignore it here and it'll get fixed when that one gets merged.

![image](https://user-images.githubusercontent.com/1522856/69997851-53c45280-1555-11ea-824a-45e4a6892c3f.png) 
![image](https://user-images.githubusercontent.com/1522856/69997867-63439b80-1555-11ea-8237-e2fc5332bec9.png)

**Note:** #1621 must be merged before this!

0nko committed:
Add a number picker dialog

0nko committed:
Add new select all button title

0nko committed:
Move all refunds-related modules into a single file

0nko committed:
Move products-related modules into a single file

0nko committed:
Move reviews-related modules into a single file

0nko committed:
Add an app model for order address

0nko committed:
Remove redundant methods

0nko committed:
Add a refund product item description string

0nko committed:
Add a view state property for the new refund by items screen

0nko committed:
Add a refund item adapter and layout

0nko committed:
Add a refund by items fragment and layout

0nko committed:
Make select all button clickable

0nko committed:
Optimize imports

0nko committed:
Tie the number picker to the refund item quantity

0nko committed:
Add refund qunatity to the list change comparison

0nko committed:
Reset refund by items state before starting a refund flow

0nko committed:
Tie the select all button to the ViewModel

0nko committed:
Display the selected items count

0nko committed:
Update refund items layout

0nko committed:
Show the refund by items fragment by default

0nko committed:
Merge branch '0nko/savedstate-step-6' into issue/1241-refund-by-items

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/IssueRefundFragment.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/IssueRefundViewModel.kt

0nko committed:
Update screen title based on selected items

0nko committed:
Include refund by items in the refund total calculation

0nko committed:
Add a tab layout with viewpager for both types of refunds

0nko committed:
Hide the toolbar shadow when tabs are visible

0nko committed:
Put a separate Next button for each fragment

0nko committed:
Reorganize the refund layouts to make them scrollable

0nko committed:
Use a child fragment when showing the tab contents

0nko committed:
Merge branch '0nko/savedstate-step-7' into issue/1241-refund-by-items

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/IssueRefundFragment.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/IssueRefundViewModel.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/RefundsModule.kt

0nko committed:
Fix the ViewModel state reset after leaving the refund flow

0nko committed:
Hide the currency intput field cursor and show/hide the keyboard based on current tab

0nko committed:
Add a refunds by items module

0nko committed:
Update the refund amount based on the refund type

0nko committed:
Add a selected total property and dynamically calculate the value based on selected items

0nko committed:
Merge branch '0nko/savedstate-step-7' into issue/1241-refund-by-items

0nko committed:
Merge branch '0nko/savedstate-step-7' into issue/1241-refund-by-items

0nko committed:
Update FluxC reference with refund by items changes

0nko committed:
Fix Snackbar event reference

0nko committed:
Add string resources

0nko committed:
Add subtotal, taxes, discounts and total refund section to the layout

0nko committed:
Add subtotal, taxes, discount and refund calculation and observers

0nko committed:
Merge branch '0nko/savedstate-step-7' into issue/1241-refund-by-items

# Conflicts:
#	build.gradle

0nko committed:
Merge branch '0nko/upgrade-to-gradle-5' into issue/1241-refund-by-items

0nko committed:
Update the navigation library to 2.1 and set the JVM target to 1.8

0nko committed:
Update the SavedState library to 1.0.0-rc02

0nko committed:
Merge branch '0nko/savedstate-step-7' into issue/1241-refund-by-items

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/IssueRefundViewModel.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/RefundSummaryFragment.kt

0nko committed:
Merge branch '0nko/upgrade-to-gradle-5' into issue/1241-refund-by-items

0nko committed:
Use the latest arch components and navigation components version

0nko committed:
Update the android annotations version

0nko committed:
Add an injector to the base fragment so that it can be used instead of DaggerFragment

0nko committed:
Use the navGraph ViewModel scope in refund fragments

0nko committed:
Provide a common navGraph SavedState owner for all refund fragments

0nko committed:
Update minor dependency versions

0nko committed:
Remove unnecessary code

0nko committed:
Remove redundant injections (the base class already does this)

0nko committed:
Merge branch 'feature/refund-by-items-master' into 0nko/number-picker-dialog

0nko committed:
Merge branch 'feature/refund-by-items-master' into 0nko/refund-items-list

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/IssueRefundViewModel.kt
#	build.gradle

0nko committed:
Fix ktlint & checkstyle issues

0nko committed:
Merge branch '0nko/number-picker-dialog' into 0nko/refund-items-list

0nko committed:
Fix the build

0nko committed:
Merge commit 'e859a669c8675bec83afbc775722961651c38a95' into 0nko/refund-items-list

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/IssueRefundFragment.kt

0nko committed:
Merge branch 'feature/refund-by-items-master' into 0nko/number-picker-dialog

0nko committed:
Merge branch '0nko/number-picker-dialog' into 0nko/refund-items-list

0nko committed:
Remove invalid comment

0nko committed:
Merge branch 'feature/refund-by-items-master' into 0nko/number-picker-dialog

0nko committed:
Merge branch '0nko/number-picker-dialog' into 0nko/refund-items-list

# Conflicts:
#	build.gradle

AmandaRiu commented:
NOTE: Added the label `Tracks Events` to this PR since it includes new tracks events.

AmandaRiu commented on WooCommerce/src/main/res/layout/fragment_refund_by_items.xml:
```diff
@@ -0,0 +1,211 @@
+                    android:layout_weight="1"
+                    android:text="@string/subtotal"
+                    android:textAlignment="viewStart"
+                    android:textAppearance="@style/Woo.OrderDetail.TextAppearance"/>
```
np: `Woo.TextAppearance`, `Woo.TextAppearance.Bold`, etc... styles should be used if not creating a custom versions for this view.

AmandaRiu reviewed (changes requested):
@0nko First round complete 😄 I've left some comments in the code, and below are some additional notes:

> By the way, I’m not sure what’s going on with ciklint failing (ktlint passes on my machine), but the Step 3 PR is passing so just ignore it here and it’ll get fixed when that one gets merged.

Thanks for the heads up. That is weird, but I think I’ve run into that before as well. 👍

### Refund snack messaging

This part is completely unrelated to this PR specifically, but is a code change in your feature branch and would’ve already been reviewed so I thought I should bring it up. First I have a question: what is the `isEndless` property of the `ShowSnackbar` event for and where is it being handled? I looked all through the code and can’t seem to find it on my own but you are setting it to `true` in `IssueRefundViewModel` and I can’t figure out what for or if it’s working. 

The issue I found in testing is if I cancel the refund request, I still see the second snackbar message. It looks like you need to move the event trigger down to be ***inside*** the `if (!wasRefundCanceled)` block:

```
val wasRefundCanceled = waitForCancellation()
triggerEvent(ShowSnackbar(
        R.string.order_refunds_amount_refund_progress_message,
        arrayOf(formatCurrency(refundByAmountState.enteredAmount)),
        isEndless = true)
)

if (!wasRefundCanceled) {
	// Trigger the event here instead
} 
```

This is just my opinion, but first showing a snackbar message with “Your refund for %s is being processed. Please wait…” with the UNDO button, and then having that disappear and be replaced by a new snackbar with exactly the same message sans the UNDO button feels really awkward to me. WDYT about changing that second message to something like “Waiting for refund confirmation”?

## Design Review
### Refund by Items View
It looks like some padding needs to be added below the ::Next:: button on this screen. You don’t really notice it in portrait mode, but in landscape mode it feels very crunched:

<img src="https://user-images.githubusercontent.com/5810477/70291781-75158080-1799-11ea-9e2f-0f6d86dc1e14.png" width="600"/>

The **Product refund** total amount doesn’t align with the right edges of the totals above it, and the button doesn’t align on theleft and  right edges with the rest of the items. The total amount color doesn’t match the design which shows it should be #3C3C3C:

![Screenshot_1575598560](https://user-images.githubusercontent.com/5810477/70291822-91b1b880-1799-11ea-9e1a-b7420173ddca.png)

### Number Picker Dialog

![Screenshot_1575599710](https://user-images.githubusercontent.com/5810477/70291825-97a79980-1799-11ea-8e25-6fc061909afa.png)

* I recommend removing the need for clicking “OK” or cancel since only a single item may be selected and changing the value doesn't require the user confirm the change. [Per the material design guidelines](https://material.io/components/dialogs/#simple-dialog) you would use a single-click Simple dialog in this scenario. The docs also recommend using a popup contextual menu instead of a dialog since it’s less disruptive and isn’t modal. If the user clicks outside of the menu it just dismisses. 

![Screen Shot 2019-12-05 at 7 54 33 PM](https://user-images.githubusercontent.com/5810477/70291831-9e361100-1799-11ea-846d-3e3d4cc482de.png)


* If a title is needed I recommend something more descriptive like “Select  quantity”. WDYT?

0nko committed:
Merge branch 'feature/refund-by-items-master' into 0nko/refund-items-list

# Conflicts:
#	WooCommerce/build.gradle
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/IssueRefundViewModel.kt

0nko committed:
Update FluxC reference to the latest revision

0nko committed:
Use let instead of a null check

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/base/BaseFragment.kt:
```diff
@@ -2,17 +2,27 @@ package com.woocommerce.android.ui.base
  * of setting the activity title
  */
-abstract class BaseFragment : androidx.fragment.app.Fragment(), BaseFragmentView {
+abstract class BaseFragment : Fragment(), BaseFragmentView, HasSupportFragmentInjector {
```
> Why remove use of DaggerFragment?

`DaggerFragment` is just a convenience class provided by Dagger that implements the `inject()` so you don't have to do it yourself. However, it injects a `Fragment` in `onAttach()`, which is too early for the way SavedState is used to provide arguments. So here we need to implement the interface ourselves. 

> Would it be a better solution would be to extend BaseFragment to create a BaseDaggerFragment for classes that require this functionality?

We could, but almost every fragment in the app needs to be injected so I'm not sure we need a separate class.

AmandaRiu replied:
Ok, I understand now. Thanks for the clarification. 👍

0nko commented on WooCommerce/src/main/res/layout/fragment_refund_by_amount.xml:
```diff
@@ -1,39 +1,66 @@
 
-    <com.google.android.material.textfield.TextInputLayout
-        android:id="@+id/issueRefund_refundAmountInputLayout"
+    <LinearLayout
```
It has 2 children. There's also the button.

AmandaRiu replied:
Oy...not sure how I missed that. 🙈

0nko committed:
Move next button tap tracking into a handler

0nko committed:
Fix the activity extensions version

0nko committed:
Update gradle plugin to version 3.5.3

0nko committed:
Update the text style references to the refunds style

0nko committed:
Remove unused parameter and don't show a snackbar if the refund is cancelled

0nko committed:
Add bottom margin to the Next button

0nko committed:
Fix the refunds button paddings

0nko commented on WooCommerce/src/main/res/layout/fragment_refund_by_items.xml:
```diff
@@ -0,0 +1,211 @@
+                    app:layout_constraintStart_toStartOf="parent"
+                    app:layout_constraintTop_toBottomOf="@id/issueRefund_selectAllButton"
+                    tools:itemCount="3"
+                    tools:listitem="@layout/refunds_product_list_item"
```
:)

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/IssueRefundViewModel.kt:
```diff
@@ -62,19 +67,21 @@ class IssueRefundViewModel @AssistedInject constructor(
-    final val refundByAmountStateLiveData = LiveDataDelegate(savedState, RefundByAmountViewState()) {
-        updateScreenTitle()
-    }
+    final val refundByItemsStateLiveData = LiveDataDelegate(savedState, RefundByItemsViewState(), onChange = {
```
Thanks!

0nko committed:
Flatten the view hierarchy

0nko commented on WooCommerce/src/main/res/layout/fragment_refund_by_items.xml:
```diff
@@ -0,0 +1,211 @@
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content">
+
+    <LinearLayout
```
I've flattened it a bit but it can't be all converted into a single `ConstraintLayout` because there's a card style that must be applied partially. Also, some `LinearLayout` are sections that can be show/hidden.

AmandaRiu replied:
> Also, some LinearLayout are sections that can be show/hidden.

You can actually do this in `ConstraintLayout` using `Group`. Just replace that `LinearLayout` with this:

```
<androidx.constraintlayout.widget.Group
    android:id="@+id/issueRefund_discountSection"
    android:layout_width="match_parent"
    android:layout_height="wrap_content"
    android:visibility="gone"
    tools:visibility="visible"
    app:constraint_referenced_ids="issueRefund_lblDiscount,issueRefund_discountItems,issueRefund_discountTotal"/>

<TextView
    android:id="@+id/issueRefund_lblDiscount"
    android:layout_width="wrap_content"
    android:layout_height="wrap_content"
    android:layout_marginStart="@dimen/card_padding_start"
    android:layout_marginTop="@dimen/card_item_padding_intra_h"
    android:text="@string/discount"
    android:textAppearance="@style/Woo.Refunds.TextAppearance"
    app:layout_constraintEnd_toStartOf="@+id/issueRefund_discountItems"
    app:layout_constraintStart_toStartOf="parent"
    app:layout_constraintTop_toBottomOf="@id/issueRefund_lblSubtotal" />

<TextView
    android:id="@+id/issueRefund_discountItems"
    android:layout_width="262dp"
    android:layout_height="wrap_content"
    android:textAlignment="viewStart"
    android:layout_marginStart="@dimen/card_item_padding_intra_h"
    android:layout_marginEnd="@dimen/card_item_padding_intra_h"
    android:textAppearance="@style/Woo.Refunds.TextAppearance"
    app:layout_constraintEnd_toStartOf="@+id/issueRefund_discountTotal"
    app:layout_constraintStart_toEndOf="@+id/issueRefund_lblDiscount"
    app:layout_constraintTop_toTopOf="@+id/issueRefund_lblDiscount"
    tools:text="(sale4theday, firstTime, anyothertime, blah)" />

<TextView
    android:id="@+id/issueRefund_discountTotal"
    android:layout_width="wrap_content"
    android:layout_height="wrap_content"
    android:layout_marginEnd="@dimen/card_padding_end"
    android:textAppearance="@style/Woo.Refunds.TextAppearance"
    app:layout_constraintEnd_toEndOf="parent"
    app:layout_constraintStart_toEndOf="@+id/issueRefund_discountItems"
    app:layout_constraintTop_toTopOf="@+id/issueRefund_lblDiscount"
    tools:text="-$4.00" />
```

> I've flattened it a bit but it can't be all converted into a single ConstraintLayout because there's a card style that must be applied partially. 

This one is a little tricker, but it seems like it is possible by using a `View`, setting the background you want to it, and then arranging all the items over it, but that seems like a lot of work and one extra layer shouldn't be that big of a performance hit.

0nko replied:
Interesting, TIL. Thanks!

0nko committed:
Update number picker dialog header

0nko committed:
Optimize imports

0nko commented:
Thanks for the review @AmandaRiu! I've addressed your comments in code and some answers to your questions here:

> what is the isEndless property of the ShowSnackbar event for and where is it being handled? 

It was supposed to be a flag for showing an indefinite snackbar but it's not really useful so I removed it.

> It looks like some padding needs to be added below the ::Next:: button on this screen. You don’t really notice it in portrait mode, but in landscape mode it feels very crunched

Fixed.

> The Product refund total amount doesn’t align with the right edges of the totals above it, and the button doesn’t align on the left and right edges with the rest of the items. 

Fixed.

> The total amount color doesn’t match the design which shows it should be #3C3C3C:

Please ignore this for now. The amount text view is actually a button from the original design, which we decided to remove. The actual amount dialog is implemented in one of the following PRs and eventually disabled in the last one. So the correct color can be verified then.

> The docs also recommend using a popup contextual menu instead of a dialog since it’s less disruptive and isn’t modal. If the user clicks outside of the menu it just dismisses.

I'm not sure a number picker view inside a context pop up menu would work. It's scrollable and the pop up view would need to be large enough so that it'd be easy to change the value. I guess the whole thing could be converted into a list but if there were a lot of items it could become very long. We can discuss it further if you like but if do decide to change it, let's do it in a separate PR. What do you think?

Ready for another round of review :).

AmandaRiu commented:
@0nko finished another round. Made a comment in the code about `ConstraintLayout`. Also the following:

* When processing a refund that has a discount, the discount section is hidden until the user selects an item for refund. Is this intentional?
* I don’t see the option to refund shipping that were in the designs. Is this intentional?

![refund-shipping](https://user-images.githubusercontent.com/5810477/70591795-90ff9480-1b94-11ea-9473-a9970a36baf4.png)

* The parentheses around the discount items from the design are missing 
* Should the **select all** button convert to a **select none** option once clicked?

0nko committed:
Replace LinearLayout with a Group

0nko committed:
Always show a discount in the refund summary when the order has one

0nko committed:
Remove discounts from refunds

0nko committed:
Show discount total as negative

0nko committed:
Optimize imports

0nko commented:
> When processing a refund that has a discount, the discount section is hidden until the user selects an item for refund. Is this intentional?

Sort of, because not all orders contain discounts. But after thinking about it, it doesn't make sense to show discounts in refunds because items already have the discount included in the price -- this is done by the API (for example, if you have an order of 2 items for $10 each and a discount for $2, the actual price for an item in the refunds would show $9). The user might be confused why the discount amount isn't applied in the total summary, so I removed that line altogether.

> I don’t see the option to refund shipping that were in the designs. Is this intentional?

Yeah. We've discovered the API doesn't support shipping refunds. But I already implemented the feature by that time. Because of this, #1623 does show it (a stub) so that it can be reviewed. It's then hidden in the next PR. I kept the code so that it can be used once the API supports it.

> Should the select all button convert to a select none option once clicked?

I've implemented this in #1633 to avoid unnecessary refactoring.

0nko committed:
Merge branch 'feature/refund-by-items-master' into 0nko/refund-items-list

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/RefundSummaryFragment.kt
#	build.gradle

AmandaRiu commented:
> The user might be confused why the discount amount isn't applied in the total summary, so I removed that line altogether.

I totally agree. It confused me. I was referring to the discount being hidden even for an order that has a discount applied - it unhides itself _after_ you select an item for a refund.

> Yeah. We've discovered the API doesn't support shipping refunds. But I already implemented the feature by that time. Because of this, #1623 does show it (a stub) so that it can be reviewed. It's then hidden in the next PR. I kept the code so that it can be used once the API supports it.

👍 

> I've implemented this in #1633 to avoid unnecessary refactoring.

👍

AmandaRiu commented:
> But after thinking about it, it doesn't make sense to show discounts in refunds because items already have the discount included in the price -- this is done by the API (for example, if you have an order of 2 items for $10 each and a discount for $2, the actual price for an item in the refunds would show $9). The user might be confused why the discount amount isn't applied in the total summary, so I removed that line altogether.

NM - I see you have this removed now.

AmandaRiu reviewed (approved):
LGTM! Nice work @0nko :shipit:
